uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/Tree.java

```diff
@@ -67,35 +73,57 @@ public void clear() {
         container.clear();
     }
 
-    void fireStateChanged(final TreeItem item,
-                          final TreeItem.State state) {
-        if (state.equals(TreeItem.State.OPEN)) {
-            OpenEvent.fire(this,
-                           item);
-        } else {
-            CloseEvent.fire(this,
-                            item);
-        }
-        onSelection(item,
-                    true);
-    }
-
-    public void setSelectedItem(final TreeItem item,
+    public void setSelectedItem(final T item,
                                 final boolean fireEvents) {
         onSelection(item,
                     fireEvents);
     }
 
-    public TreeItem getSelectedItem() {
+    public T getSelectedItem() {
         return curSelection;
     }
 
-    public void setSelectedItem(final TreeItem item) {
+    public void setSelectedItem(final T item) {
         onSelection(item,
                     true);
     }
 
-    void onSelection(final TreeItem item,
+    @SuppressWarnings("unchecked")
+    public T addItem(final T item) {
+        container.add(item);
+        item.setTree((Tree<TreeItem>) this);
+        return item;
+    }
+
+    public T getItem(int itemId) {
```

Comment:
the argument's name is not clear.. it's called `idemId`, but in fact it's the item index, right? I would change it to `itemIndex` or just `index`...